membership/src/main/scala/zio/membership/hyparview/InitialProtocol.scala

```diff
@@ -0,0 +1,76 @@
+package zio.membership.hyparview
+
+import upickle.default._
+import zio.membership.ByteCodec
+
+sealed abstract class InitialProtocol[T]
+
+object InitialProtocol {
+
+  implicit def tagged[T](
+    implicit
+    c1: ByteCodec[Neighbor[T]],
+    c2: ByteCodec[Join[T]],
+    c3: ByteCodec[ForwardJoinReply[T]],
+    c4: ByteCodec[ShuffleReply[T]]
+  ): TaggedCodec[InitialProtocol[T]] =
+    TaggedCodec.instance(
+      {
+        case _: Neighbor[T]         => 0
+        case _: Join[T]             => 1
+        case _: ForwardJoinReply[T] => 2
+        case _: ShuffleReply[T]     => 3
+      }, {
+        case 0 => c1.asInstanceOf[ByteCodec[InitialProtocol[T]]]
```

Comment:
Please create an issue to document these, either directly in code or in a separate document.

Reply:
They will hopefully become obsolete once we use a more powerful codec.
In the meantime I would say they are not really interesting unless you intend to look at the raw messages

Reply:
we need something for that. it looks ugly and it will be hard to maintain and debug.

I was thinking if we could add all `ByteCodec` into array and use index as tag.  
then interface will look like that.
`TaggedCodec.make[ForwardJoinReply, ShuffleReply, ...]` 

@mschuwalow what do you think? 

@mschuwalow @mijicd I will open PR to move `TaggedCodec` and `ByteCodec` to core because it will be need for SWIM refactor anyway.